File: frontend/src/types/fleet_matrix.ts
import type {
  ComplianceStatus,
  DeviceCategory,
  HealthRating,
  KeyExpiryStatus,
  SSLCertificateStatus,
  TailnetLockStatus,
} from './enums';

export interface FleetMatrixColumn {
  key: string;
  label: string;
  category: string;
  type: 'string' | 'boolean' | 'number' | 'list' | 'datetime' | 'status';
  sortable: boolean;
  filterable: boolean;
}

export interface NodeMatrixItem {
  id: string;
  node_id: string | null;
  name: string;
  hostname: string;
  user: string | null;
  tailnet: string | null;
  addresses: string[];
  tags: string[];

  // System & OS
  os: string;
  os_raw: string | null;
  os_version: string | null;
  client_version: string | null;
  category: DeviceCategory | string;

  // Status & Uptime
  is_online: boolean;
  last_seen: string | null;
  streak_duration_seconds: number;
  streak_duration_human: string;

  // Security Posture & Compliance
  is_compliant: boolean;
  compliance_status: ComplianceStatus;
  violations: string[];
  update_available: boolean;

  // Key Lifecycle
  key_expiry_disabled: boolean;
  expires_at: string | null;
  key_expiry_status: KeyExpiryStatus;
  days_until_key_expiry: number | null;

  // Network & Routing
  is_exit_node: boolean;
  exposed_routes: string[];
  derp_region: string | null;
  latency_ms: number | null;

  // Geolocation
  country: string | null;
  public_address: string | null;
  country_name?: string | null;
  is_impossible_travel?: boolean;
  geo_anomaly_reason?: string | null;

  // Tailnet Lock
  tailnet_lock_status: TailnetLockStatus;
  is_locked_out: boolean;
  is_quarantined: boolean;
  is_signing_node: boolean;

  // MagicDNS & TLS
  magicdns_domain: string | null;
  ssl_cert_status: SSLCertificateStatus;
  ssl_cert_expires_at: string | null;
  is_ssl_cert_expiring: boolean;
  is_ssl_cert_expired: boolean;

  // Overall Node Health
  health_status: HealthRating;
  health_score: number;

  [key: string]: unknown;
}

export interface FleetMatrixCrossTabulation {
  os_by_status: Record<string, Record<string, number>>;
  os_by_compliance: Record<string, Record<string, number>>;
  category_by_status: Record<string, Record<string, number>>;
}

export interface FleetMatrixResponse {
  total_devices: number;
  filtered_devices: number;
  limit: number;
  offset: number;
  has_more: boolean;
  columns: FleetMatrixColumn[];
  cross_tabulation: FleetMatrixCrossTabulation;
  matrix: NodeMatrixItem[];
  timestamp: string;
}

export interface FleetMatrixSummaryResponse {
  total_devices: number;
  online_devices: number;
  offline_devices: number;
  cross_tabulation: FleetMatrixCrossTabulation;
  columns: FleetMatrixColumn[];
  timestamp: string;
}

export interface FleetMatrixQueryParams {
  tailnet?: string;
  status?: 'online' | 'offline' | 'all';
  os?: string;
  compliance?: 'compliant' | 'non_compliant';
  category?: DeviceCategory | string;
  locked_out?: boolean;
  quarantined?: boolean;
  geo_anomaly?: boolean;
  exit_node?: boolean;
  q?: string;
  sort_by?: string;
  order?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

File: backend/app/schemas/fleet_matrix.py
from __future__ import annotations

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class FleetMatrixColumn(BaseModel):
    """Schema describing column attributes and UI metadata for the fleet matrix."""

    key: str
    label: str
    category: str
    type: str  # 'string', 'boolean', 'number', 'list', 'datetime', 'status'
    sortable: bool = True
    filterable: bool = True

    model_config = ConfigDict(populate_by_name=True)


class NodeMatrixItem(BaseModel):
    """Unified multidimensional matrix row for a Tailscale node across all telemetry dimensions."""

    id: str
    node_id: Optional[str] = None
    name: str
    hostname: str
    user: Optional[str] = None
    tailnet: Optional[str] = None
    addresses: List[str] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)

    # System & OS
    os: str
    os_raw: Optional[str] = None
    os_version: Optional[str] = None
    client_version: Optional[str] = None
    category: str = "unknown"

    # Status & Uptime
    is_online: bool = False
    last_seen: Optional[str] = None
    streak_duration_seconds: float = 0.0
    streak_duration_human: str = "< 1m"

    # Security Posture & Compliance
    is_compliant: bool = True
    compliance_status: str = "compliant"
    violations: List[str] = Field(default_factory=list)
    update_available: bool = False

    # Key Lifecycle
    key_expiry_disabled: bool = False
    expires_at: Optional[str] = None
    key_expiry_status: str = "healthy"
    days_until_key_expiry: Optional[float] = None

    # Network & Routing
    is_exit_node: bool = False
    exposed_routes: List[str] = Field(default_factory=list)
    derp_region: Optional[str] = None
    latency_ms: Optional[float] = None

    # Geolocation
    country: Optional[str] = None
    public_address: Optional[str] = None
    country_name: Optional[str] = None
    is_impossible_travel: bool = False
    geo_anomaly_reason: Optional[str] = None

    # Tailnet Lock
    tailnet_lock_status: str = "unknown"
    is_locked_out: bool = False
    is_quarantined: bool = False
    is_signing_node: bool = False

    # MagicDNS & TLS
    magicdns_domain: Optional[str] = None
    ssl_cert_status: str = "not_configured"
    ssl_cert_expires_at: Optional[str] = None
    is_ssl_cert_expiring: bool = False
    is_ssl_cert_expired: bool = False

    # Overall Node Health
    health_status: str = "healthy"  # 'healthy', 'warning', 'critical'
    health_score: float = 100.0

    model_config = ConfigDict(populate_by_name=True, extra="allow")


class FleetMatrixCrossTabulation(BaseModel):
    """Cross-tabulated 2D summary matrix breakdown across fleet dimensions."""

    os_by_status: Dict[str, Dict[str, int]] = Field(default_factory=dict)
    os_by_compliance: Dict[str, Dict[str, int]] = Field(default_factory=dict)
    category_by_status: Dict[str, Dict[str, int]] = Field(default_factory=dict)

    model_config = ConfigDict(populate_by_name=True)


class FleetMatrixResponse(BaseModel):
    """Full API response for the fleet matrix endpoint with pagination, columns, and matrix rows."""

    total_devices: int
    filtered_devices: int
    limit: int
    offset: int
    has_more: bool
    columns: List[FleetMatrixColumn] = Field(default_factory=list)
    cross_tabulation: FleetMatrixCrossTabulation = Field(default_factory=FleetMatrixCrossTabulation)
    matrix: List[NodeMatrixItem] = Field(default_factory=list)
    timestamp: str

    model_config = ConfigDict(populate_by_name=True)
